src/server/replication_messages.rs

```diff
@@ -175,36 +427,123 @@ impl InitMessage {
+    fn reset(&mut self) {
+        self.cursor.set_position(0);
         self.entities.clear();
-        self.buffer.reset();
+    }
+
+    /// Starts writing entity and its data.
+    ///
+    /// Data can contain components with their IDs.
+    /// Entity will be written lazily after first data write.
+    /// See also [`Self::end_entity_data`] and [`Self::write_component`].
+    pub(super) fn start_entity_data(&mut self, entity: Entity) {
+        debug_assert_eq!(self.entity_data_size, 0);
+
+        self.data_entity = entity;
+        self.entity_data_pos = self.cursor.position();
+    }
+
+    /// Writes entity for the current data and remembers the position after it to write length later.
+    ///
+    /// Should be called only after first data write.
+    fn write_data_entity(&mut self) -> bincode::Result<()> {
+        serialize_entity(&mut self.cursor, self.data_entity)?;
+        self.entity_data_size_pos = self.cursor.position();
+        self.cursor.set_position(
+            self.entity_data_size_pos + mem::size_of_val(&self.entity_data_size) as u64,
+        );
 
         Ok(())
     }
 
-    /// Registers entity from buffer's entity data and its size for possible splitting.
-    pub(super) fn register_entity(&mut self) {
-        let data_size = self.buffer.as_slice().len() - self.buffer.entity_data_pos() as usize;
-        self.entities.push((self.buffer.data_entity(), data_size));
+    /// Ends writing entity data by writing its length into the last remembered position.
+    ///
+    /// If the entity data is empty, nothing will be written.
+    /// See also [`Self::start_array`] and [`Self::write_component`].
+    pub(super) fn end_entity_data(&mut self) -> bincode::Result<()> {
+        if self.entity_data_size == 0 {
+            self.cursor.set_position(self.entity_data_pos);
+            return Ok(());
+        }
+
+        if self.entity_data_size == 0 {
+            self.write_data_entity()?;
+        }
+
+        let previous_pos = self.cursor.position();
+        self.cursor.set_position(self.entity_data_size_pos);
+
+        bincode::serialize_into(&mut self.cursor, &self.entity_data_size)?;
+
+        self.cursor.set_position(previous_pos);
+
+        let data_size = self.cursor.position() - self.entity_data_pos;
+        self.entities.push((self.data_entity, data_size as usize));
+
+        self.entity_data_size = 0;
+
+        Ok(())
     }
 
-    /// Returns `true` is message contains any written data.
-    fn is_sendable(&self) -> bool {
-        !self.buffer.as_slice().is_empty()
+    /// Serializes component and its replication ID as an element of entity data.
+    ///
+    /// Should be called only inside an entity data and increases its size.
+    /// See also [`Self::start_entity_data`].
+    pub(super) fn write_component(
+        &mut self,
+        replication_info: &ReplicationInfo,
+        replication_id: ReplicationId,
+        ptr: Ptr,
+    ) -> bincode::Result<()> {
```

Comment:
Not a fan either, but what would you suggest to avoid duplication? They use many fields internally.
I guess I can only put this part:
```rust
        let previous_pos = self.cursor.position();
        DefaultOptions::new().serialize_into(&mut self.cursor, &replication_id)?;
        (replication_info.serialize)(ptr, &mut self.cursor)?;

        let component_size = (self.cursor.position() - previous_pos)
            .try_into()
            .map_err(|_| bincode::ErrorKind::SizeLimit)?;
```
into a function that returns size, what do you think?